src/oidcendpoint/oidc/token.py

```diff
@@ -111,50 +254,132 @@ def _access_token(self, req, **kwargs):
-                return self.error_cls(error="invalid_request")
+        try:
+            keyjar = self.endpoint_context.keyjar
+        except AttributeError:
+            keyjar = ""
 
-        if "client_id" not in request:  # Optional for access token request
-            request["client_id"] = client_id
+        request.verify(keyjar=keyjar, opponent_id=client_id)
 
-        logger.debug("%s: %s" % (request.__class__.__name__, sanitize(request)))
+        _mngr = self.endpoint_context.session_manager
+        try:
+            _session_info = _mngr.get_session_info_by_token(request["refresh_token"])
+        except KeyError:
+            logger.error("Access Code invalid")
+            return self.error_cls(error="invalid_grant")
+
+        token = _mngr.find_token(_session_info["session_id"], request["refresh_token"])
+
+        if not isinstance(token, RefreshToken):
+            return self.error_cls(
+                error="invalid_request", error_description="Wrong token type"
+            )
+
+        if token.is_active() is False:
+            return self.error_cls(
+                error="invalid_request", error_description="Refresh token inactive"
+            )
 
         return request
 
-    def process_request(self, request=None, **kwargs):
+
+HELPER_BY_GRANT_TYPE = {
+    "authorization_code": AccessTokenHelper,
+    "refresh_token": RefreshTokenHelper,
+}
+
+
+class Token(Endpoint):
+    request_cls = oidc.Message
+    response_cls = oidc.AccessTokenResponse
+    error_cls = TokenErrorResponse
+    request_format = "json"
+    request_placement = "body"
+    response_format = "json"
+    response_placement = "body"
+    endpoint_name = "token_endpoint"
+    name = "token"
+    default_capabilities = {"token_endpoint_auth_signing_alg_values_supported": None}
+
+    def __init__(self, endpoint_context, new_refresh_token=False, **kwargs):
+        Endpoint.__init__(self, endpoint_context, **kwargs)
+        self.post_parse_request.append(self._post_parse_request)
+        if "client_authn_method" in kwargs:
+            self.endpoint_info["token_endpoint_auth_methods_supported"] = kwargs[
+                "client_authn_method"
+            ]
+        self.allow_refresh = False
+        self.new_refresh_token = new_refresh_token
+        self.configure_grant_types(kwargs.get("grant_types_supported"))
+
+    def configure_grant_types(self, grant_types_supported):
+        if grant_types_supported is None:
+            self.helper = {k: v(self) for k, v in HELPER_BY_GRANT_TYPE.items()}
+            return
+
+        self.helper = {}
+        # TODO: do we want to allow any grant_type?
+        for grant_type, grant_type_options in grant_types_supported.items():
+            if (
+                    grant_type_options in (None, True)
+                    and grant_type in HELPER_BY_GRANT_TYPE
+            ):
+                self.helper[grant_type] = HELPER_BY_GRANT_TYPE[grant_type]
```

Comment:
The class will need to be initialized ->
`self.helper[grant_type] = HELPER_BY_GRANT_TYPE[grant_type](self)`